rts/idris_rts.c

```diff
@@ -272,6 +286,22 @@ char* GETSTROFF(VAL stroff) {
     return (root->str->info.str + root->offset);
 }
 
+VAL MKCDATA(VM* vm, CHeapItem * item) {
+    Closure* cl = allocate(sizeof(Closure), 0);
+    SETTY(cl, CDATA);
+    cl->info.c_heap_item = item;
+    c_heap_insert_if_needed(vm, &vm->c_heap, item);
+    return cl;
+}
+
+VAL MKCDATAc(VM* vm, CHeapItem * item) {
+    Closure* cl = allocate(sizeof(Closure), 1);
+    SETTY(cl, CDATA);
+    cl->info.c_heap_item = item;
+    c_heap_insert_if_needed(vm, &vm->c_heap, item);
+    return cl;
```

Comment:
The heap insert may trigger a gc and cl wouldn't be a valid pointer after that. Simplest is to allocate the closure after the heap insert is done.